### PR TITLE
Prefer mold or lld as linker to speed up linking

### PR DESCRIPTION
**Pull request by daschuer**
This uses mold or lld as linker if available. 

It also installs it on our Ubuntu CI. The release builds on Launchpad are still on ld. 
So the 2.4 merge target should be fine. 

We may also consider to use lld for macOS, but that would effect our release builds. 
Not sure about the consequences. It seems to be used by Google for Chrome. 
`mold` is not available on macOS requires the commercial clone `sold` 

Discussed here: 
https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Reducing.20linking.20times

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -541,6 +541,46 @@ else()
   message(STATUS "Support for ccache: ${CCACHE_SUPPORT}")
 endif()
 
+if(NOT MSVC)
+  find_program(MOLD_EXECUTABLE mold)
+  default_option(MOLD_SUPPORT "Enable ccache support" "MOLD_EXECUTABLE")
```
Comment (Holzhaus) [inline CMakeLists.txt:546]:
```suggestion
  default_option(MOLD_SUPPORT "Use 'mold' for linking" "MOLD_EXECUTABLE")
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -541,6 +541,46 @@ else()
...
+  else()
+    find_program(LLD_EXECUTABLE lld)
+    if(LLD_EXECUTABLE)
+      execute_process(
+        COMMAND ${CMAKE_CXX_COMPILER} -fuse-ld=lld -Wl,--version
+        ERROR_QUIET
+        OUTPUT_VARIABLE LLD_VERSION_STRING
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+      )
+      string(REGEX MATCH "LLD ([0-9]+\\.[0-9]+\\.[0-9]+)" LLD_VERSION_MATCH ${LLD_VERSION_STRING})
```
Comment (Holzhaus) [inline CMakeLists.txt:563]:
```suggestion
      string(REGEX MATCH "LLD ([0-9]+\\.[0-9]+\\.[0-9]+)" LLD_VERSION_MATCH "${LLD_VERSION_STRING}")
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -541,6 +541,46 @@ else()
...
+      if(LLD_VERSION_MATCH)
+        set(LLD_VERSION ${CMAKE_MATCH_1})
+        message(STATUS "Found lld with version: ${LLD_VERSION}")
+      else()
+        message(WARNING "Failed to parse lld version from: ${LLD_VERSION_STRING}")
+      endif()
+    endif()
+    # LLD 10.0.0 does not work because of https://bugs.llvm.org/show_bug.cgi?id=45769
+    default_option(LLD_SUPPORT "Enable ccache support" "LLD_EXECUTABLE;LLD_VERSION VERSION_GREATER 10.0.0")
```
Comment (Holzhaus) [inline CMakeLists.txt:572]:
```suggestion
    default_option(LLD_SUPPORT "Use 'lld' for linking" "LLD_EXECUTABLE;LLD_VERSION VERSION_GREATER 10.0.0")
```

File: tools/debian_buildenv.sh [hunk 4]
```diff
@@ -34,6 +34,16 @@ case "$1" in
             sudo apt-get install libjack-jackd2-dev;
         fi
 
+        # Install a faster linker. Prefer mold, fall back to lld
+        if apt-cache show mold 2>%1 >/dev/null;
+        then
+            sudo apt-get install mold
```
Comment (Holzhaus) [inline tools/debian_buildenv.sh:40]:
```suggestion
            sudo apt-get install -y --no-install-recommends mold
```

File: tools/debian_buildenv.sh [hunk 5]
```diff
@@ -34,6 +34,16 @@ case "$1" in
...
+        else
+            if apt-cache show lld 2>%1 >/dev/null;
+            then
+                sudo apt-get install lld
```
Comment (Holzhaus) [inline tools/debian_buildenv.sh:44]:
```suggestion
                sudo apt-get install -y --no-install-recommends lld
```

File: tools/debian_buildenv_qt6.sh [hunk 6]
```diff
@@ -34,6 +34,16 @@ case "$1" in
             sudo apt-get install libjack-jackd2-dev;
         fi
 
+        # Install a faster linker. Prefer mold, fall back to lld
+        if apt-cache show mold 2>%1 >/dev/null;
+        then
+            sudo apt-get install mold
```
Comment (Holzhaus) [inline tools/debian_buildenv_qt6.sh:40]:
```suggestion
            sudo apt-get install -y --no-install-recommends mold
```

File: tools/debian_buildenv_qt6.sh [hunk 7]
```diff
@@ -34,6 +34,16 @@ case "$1" in
...
+        else
+            if apt-cache show lld 2>%1 >/dev/null;
+            then
+                sudo apt-get install lld
```
Comment (Holzhaus) [inline tools/debian_buildenv_qt6.sh:44]:
```suggestion
                sudo apt-get install -y --no-install-recommends lld
```

**Review by Holzhaus — COMMENTED** [review]
Thanks!

**Review by Holzhaus — APPROVED** [review]
LGTM, thanks.

File: CMakeLists.txt [hunk 8]
```diff
@@ -541,6 +541,46 @@ else()
   message(STATUS "Support for ccache: ${CCACHE_SUPPORT}")
 endif()
 
+if(NOT MSVC)
+  find_program(MOLD_EXECUTABLE mold)
+  default_option(MOLD_SUPPORT "Use 'mold' for linking" "MOLD_EXECUTABLE")
+  if(MOLD_SUPPORT)
+    if(MOLD_EXECUTABLE)
+      message(STATUS "Selecting mold as linker")
+      add_link_options("-fuse-ld=mold")
```
Comment (ronso0) [inline CMakeLists.txt:550]: Mold support was added in GCC 12.1.0, hence the build will fail with older versions.
I'm not sure if the minimum supported OS versions provide 12.10.0+ or if this section should be guarded by a additional gcc version check.

Comment (ronso0) [reply CMakeLists.txt:550]: alternative for GCC < 12.10.0:
`-DCMAKE_EXE_LINKER_FLAGS:STRING=-B/usr/bin/mold`
+
`-DCMAKE_LINKER:FILEPATH=/usr/bin/mold`
https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Reducing.20linking.20times/near/377606740
from https://github.com/rui314/mold#how-to-use

**Pull request merged**